fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_callback.py

```diff
@@ -0,0 +1,103 @@
+
+logger = logging.getLogger(__name__)
+
+
+class CyclicalSGLDSamplingCallback(Callback):
+    def __init__(self,
+                 n_epochs: int,
+                 n_samples: int,
+                 n_thinning: int,
+                 cycle_length: int,
+                 exploration_ratio: float,
+                 trainer: TrainerABC,
+                 state_repository: TrainStateRepository,
+                 keep_top_n_checkpoints: int,
+                 save_checkpoint_dir: Optional[Path] = None,
+                 ):
+        """
+        Cyclical Stochastic Gradient Langevin Dynamics (SGLD) callback that collects samples
+        in different cycles. See `Zhang R. et al., 2020 <https://openreview.net/pdf?id=rkeS1RVtPS>`_
+        for more details.
+
+        Parameters
+        ----------
+        n_epochs: int
+            The number of training epochs.
+        n_samples: int
+            The desired number of the posterior samples.
+        n_thinning: int
+            Keep only each `n_thinning` sample during the sampling phase.
+        cycle_length: int
+            The length of each exploration/sampling cycle, in steps.
+        init_step_size: float
+            The initial step size.
+        exploration_ratio: float
+            The fraction of steps to allocate to the mode exploration phase.
+        trainer: TrainerABC
+            An instance of the trainer class.
+        state_repository: TrainStateRepository
+            An instance of the state repository.
+        keep_top_n_checkpoints: int
+            Number of past checkpoint files to keep.
+        save_checkpoint_dir: Optional[Path]
+            The optional path to save checkpoints.
+        """
+        if n_samples * cycle_length < cycle_length != 0:
+            raise ValueError("The number of desired samples per cycle `n_samples` * `n_thinning` "
+                             f"= {n_samples * cycle_length} is less than `cycle_length` = {cycle_length}.")
+        self.n_samples = n_samples
+        self.n_thinning = n_thinning
+        self.cycle_length = cycle_length
+        self.exploration_ratio = exploration_ratio
+        self.trainer = trainer
+        self.state_repository = state_repository
+        self.keep_top_n_checkpoints = keep_top_n_checkpoints
+        self.save_checkpoint_dir = save_checkpoint_dir
+
+        self.samples_per_epoch = n_samples // n_epochs
+        self.current_step = 0
+        self.current_epoch = 0
+        self.samples_count = 0
+
+    def do_sample(self) -> bool:
+        return ((self.current_step % self.cycle_length) / self.cycle_length) < self.exploration_ratio \
+            and (self.current_step % self.cycle_length) % self.n_thinning == 0 \
+            and self.samples_count < self.n_samples
+
+    def training_step_end(self, state: TrainState) -> TrainState:
+        self.current_step += 1
+
+        if self.do_sample():
+            if self.save_checkpoint_dir:
+                self.trainer.save_checkpoint(
```

Comment:
Do you need `save_checkpoint`? It seem you are anyway putting the state in the state repository. If a `checkpoint_dir` is passed o the repository, the state is anyway saved on disk. 

Am I missing something?

Reply:
This is similar to https://github.com/awslabs/fortuna/blob/main/fortuna/prob_model/posterior/deep_ensemble/deep_ensemble_posterior.py#L106-L110 in DeepEnsembles: the state is a collection of checkpoints, which are getting saving even if `dump_state` is False. We can throw a warning about that to inform the user.